xdsl/backend/riscv/register_allocation.py

```diff
@@ -46,6 +46,23 @@ def allocate(self, reg: SSAValue) -> bool:
 
         return False
 
+    def process_operation(self, op: Operation) -> None:
+        match op:
+            case riscv_scf.ForOp():
+                self.allocate_for_loop(op)
+            case RISCVOp():
+                self.process_riscv_op(op)
+            case _:
+                return
+
+    @abc.abstractmethod
+    def process_riscv_op(self, op: RISCVOp) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def allocate_for_loop(self, loop: riscv_scf.ForOp) -> None:
+        raise NotImplementedError()
```

Comment:
Yes, I feel these kind of abstract interfaces start to add overhead and may cause confusion, e.g. because we need to agree on invariants across these different register allocators, ...

I personally would have chosen a different approach to resolve our register allocator chaos. Having said this, this patch series seems to go towards clearing up some of the chaos that we have. So let's follow through with it. I see @superlopuh has a plan, and I believe this plan can work out. (Add I hope that some of my concerns will be resolved automatically as @superlopuh completes his mission).

Reply:
I added some docstrings. This commit precedes the one that changes the walking direction of the non-livenes allocator to walk backwards, and I haven't found a clean way to swap the two changes, it feels much easier once some common functionality has been factored out.